src/main/kotlin/org/sd_jwt/SdJwt.kt

```diff
@@ -210,6 +244,72 @@ inline fun <reified T> createCredential(
     return sdJwtEncoded + SEPARATOR + disclosures.joinToString(SEPARATOR)
 }
 
+/**
+ * @suppress
+ * This method is not for API users.
+ *
+ * Verifies if keys (and sub keys) of firstJson exist in secondJson
+ *
+ */
+fun validateJSON(
+    firstJson: JSONObject,
+    secondJson: JSONObject
+): Boolean {
+    val keys = firstJson.keys()
+
+    while (keys.hasNext()) {
+        val key = keys.next() as String
+
+        if (secondJson.has(key).not()) {
+            return false
+        }
+
+        val value1 = firstJson[key]
+        val value2 = secondJson[key]
+
+        // Recursively check the structure for nested JSONObjects or JSONArrays
+        if (value1 is JSONObject && value2 is JSONObject) {
+            if (!validateJSON(value1, value2)) {
+                return false
+            }
+        } else if (value1 is JSONArray && value2 is JSONArray) {
+            if (!validateJSONArray(value1, value2)) {
+                return false
+            }
+        } else if (
+            value1 is JSONArray && value2 !is JSONArray ||
+            value1 !is JSONArray && value2 is JSONArray
+        ) {
+            // value types should match
+            return false
+        }
+    }
+
+    return true
+}
+
+/**
+ * @suppress
+ * This method is not for API users.
+ *
+ * Verifies if elements of firstArray and secondArray are matching
+ *
+ */
+fun validateJSONArray(firstArray: JSONArray, secondArray: JSONArray): Boolean {
```

Comment:
I think there should be a check that both arrays have the same number of elements. Otherwise it could happen that array items are not checked.

Reply:
hmm..I am honestly not sure how to deal with this in the case of json arrays, thats why I asked the question on how a discloseStructure can be build in the case of a json array.

In fact, similarly to the logic in `validateJSON` , the goal is to check if elements of `firstArray` should all b present in `secondArray`. Which means we are validating the disclosureStructure/releaseClaims against the original claims of the credential. Basically, the claims provided in those structures should be valid ones. Which means that the element number, whether it is a json object or a json array could be different. example:

if I have a claim in my credential as follows:
```
"complexClaim" : [
{"key1" : "value1" , "key2":"value2"},
{"key3": "value3", "key4":"value4"}
{"key5": "value5", "key6":"value6"}
]

```
how can I build the disclosure structure accordingly, if I want to selectively disclose some elements of the array e.g the second element of the array?

```
"complexClaim" : [
{},
{"key3": "", "key4":""}
]
```
-> this will disclose only the subclaims of the second element of the array
-> the first element is an empty object, so it will be ignored (no disclosure) and only used to match which element number of the array

---
Or, should the array always have the same number of elements?

```
"complexClaim" : [
{},
{"key3": "", "key4":""},
{}
]
```
In this case, it makes sense to make an additional check on the array number of elements
It would be great if you give an example to understand what you mean so that I include it as a test case.

Reply:
Right now we have in arrays only one type of objects and the disclose structure contains one array item to signal that the items in the array should be disclosed. Disclosing single items from an array is not yet supported but on the roadmap.

So what should be support is that we have the following credential:
```
"complexClaim" : [
  {"key1" : "value1" , "key2":"value2"},
  {"key1": "value3", "key2":"value4"}
  {"key1": "value5", "key2":"value6"}
]
```
And the disclose structure would be
```
"complexClaim": [
  {"key1": "", "key2": ""}
]
```
to disclose both keys in every array item.

Under this condition I have to retract the requirement for checking the length of both arrays. That makes obviously no sense.

Does this make sense? I am not sure if this is supported by the validation function right now. Let me know what you think.

Reply:
hmm I see. Actually, with the current implementation we are validating array elements pair by pair, where the array taken from the disclosureStructure can include more than 1 element (contrary to the example you provided). That means, we still can expect different element structure within the same array, since they are the same type (JSONObject) but can have completely different structure (which should not be the case here, as you mentioned in a previous old comment).
In that sense and following the example you provided, I think the function `validateJSONArray(firstArray: JSONArray, secondArray: JSONArray)` must change as follows:

1- validate that all elements of `secondArray` have exactly the same structure (e.g identical Map keys including nested Maps etc..).
2- `firstArray` must contain only 1 element
3- The following check should pass: `validateJSON(firstArray[0],secondArray[0])`
Any other case should fail the validation.

The tests should also be reviewed and adjusted, since they were passing even though the logic is not correct.

What is your opinion on that, more specifically the above 3-step algorithm?

Reply:
Yes I think your 3-step algorithm should be correct but aren't the createSdClaims and findDisclosures methods doing exactly this? Do we even need the validation before hand? Maybe I am missing something here.